cartridges/int_bolt_embedded_sfra/cartridge/client/default/js/analytics/actions.js

```diff
@@ -0,0 +1,212 @@
+/* eslint-disable no-restricted-syntax */
+
+'use strict';
+
+const click = event('click');
+
+module.exports = {
+    waitFor,
+    urlMatchParts,
+    observeSelector,
+    render,
+    event,
+    click,
+    input: event('input'),
+    change: event('change'),
+    clickOnce: event('click', true),
+    inputOnce: event('input', true),
+    areAllFieldsFilled,
+    accountRegistrationCheckbox,
+    stepChange
+};
+
+const noop = () => {};
+
+// Don't use this function for selector matching. Use observeSelector instead.
+/**
+ * Waits for a condition to be true and emits an event.
+ * @param {Function} fn - The condition function.
+ * @param {Function} [eventData=noop] - The event data function.
+ * @returns {boolean} - True if the condition is met, false otherwise.
+ */
+function waitFor(fn, eventData = noop) {
+    return (emit, context) => {
+        if (fn(context)) {
+            emit(eventData(context));
+            return true;
+        }
+        return false;
+    };
+}
+
+/**
+ * Matches URL parts and waits for the condition to be true.
+ * @param {string[]} urlSubstrings - The URL substrings to match.
+ * @param {Function} [eventData=noop] - The event data function.
+ * @param {Location} [location=window.location] - The location object.
+ * @returns {boolean} - True if the condition is met, false otherwise.
+ */
+function urlMatchParts(urlSubstrings, eventData = noop, location = window.location) {
+    const lowercase = urlSubstrings.map(url => url.toLowerCase());
+    const partsMatch = () => lowercase.every(part => location.href.toLowerCase().includes(part));
+
+    return waitFor(partsMatch, eventData);
+}
+
+/**
+ * Observes elements matching a selector and invokes a callback for each element.
+ * @param {string} selector - The CSS selector.
+ * @param {Function} callback - The callback function.
+ * @returns {boolean} - Always returns true.
+ */
+function observeSelector(selector, callback) {
+    return (emit, context) => {
+        const elements = document.querySelectorAll(selector);
+
+        for (const element of elements) {
+            callback(emit, context, element, { disconnect: () => undefined });
```

Comment:
You're thinking of the `render` action.

This `observeSelector` action will call the callback every time the DOM changes and the target selector is found. For example, if I'm observing on `.email`, then every time a new element with `.email` class name appears, it will fire the callback. We also do a call before the mutation observer in case the DOM already has elements with `.email` before we start observing.